### PR TITLE
Fix #21328

### LINKED ISSUE #21328: VecMap::into_iter shouldn't be taking a `&mut self`
cc @Gankro 

I'm in the process of `IntoIter`-ifying the collections, and I've discovered that the `VecMap::into_iter` takes a `&mut self` instead of the convention of `into_*` consuming the value. It's also marked stable, even though the type itself is marked unstable. Should this instead be a candidate for the `drain` method discussed in the collection reform trait?

### PR DESCRIPTION
**Pull request by aochagavia**
Fixes #21328 

**Breaking change**: `VecMap::into_iter` now consumes the `VecMap`. To fix it you should pass the VecMap by value instead of by reference.

[breaking-change]

r? @Gankro

### CODE REVIEW

File: src/libcollections/vec_map.rs [hunk 1]
```diff
@@ -265,10 +265,10 @@ impl<V> VecMap<V> {
     }
 
     /// Returns an iterator visiting all key-value pairs in ascending order by
-    /// the keys, emptying (but not consuming) the original `VecMap`.
+    /// the keys, consuming the original `VecMap`.
     /// The iterator's element type is `(uint, &'r V)`.
     ///
-    /// # Examples
+    /// # Example
```
Comment (Gankro) [inline src/libcollections/vec_map.rs]: Examples is the convention, regardless of number of exmaples.

File: src/libcollections/vec_map.rs [hunk 2]
```diff
@@ -284,14 +284,42 @@ impl<V> VecMap<V> {
...
             v.map(|v| (i, v))
         }
         let filter: fn((uint, Option<V>)) -> Option<(uint, V)> = filter; // coerce to fn ptr
 
-        let values = replace(&mut self.v, vec!());
-        IntoIter { iter: values.into_iter().enumerate().filter_map(filter) }
+        IntoIter { iter: self.v.into_iter().enumerate().filter_map(filter) }
+    }
+
+    /// Returns an iterator visiting all key-value pairs in ascending order by
```
Comment (Gankro) [inline src/libcollections/vec_map.rs]: I would go with "order *of* the key*

File: src/libcollections/vec_map.rs [hunk 3]
```diff
@@ -284,14 +284,42 @@ impl<V> VecMap<V> {
...
+    ///
+    /// ```
+    /// use std::collections::VecMap;
+    ///
+    /// let mut map = VecMap::new();
+    /// map.insert(1, "a");
+    /// map.insert(3, "c");
+    /// map.insert(2, "b");
+    ///
+    /// // Not possible with .iter()
```
Comment (Gankro) [inline src/libcollections/vec_map.rs]: I disagree with this claim. It's easy to map the input in this case to get the same result with `iter` in this case.

File: src/libcollections/vec_map.rs [hunk 4]
```diff
@@ -284,14 +284,42 @@ impl<V> VecMap<V> {
...
+    /// let vec: Vec<(uint, &str)> = map.drain().collect();
+    ///
+    /// assert_eq!(vec, vec![(1, "a"), (2, "b"), (3, "c")]);
+    /// ```
+    #[unstable = "matches collection reform specification, waiting for dust to settle"]
+    pub fn drain<'a>(&'a mut self) -> Drain<'a, V> {
+        fn filter<A>((i, v): (uint, Option<A>)) -> Option<(uint, A)> {
+            v.map(|v| (i, v))
+        }
+        let filter: fn((uint, Option<V>)) -> Option<(uint, V)> = filter; // coerce to fn ptr
```
Comment (Gankro) [inline src/libcollections/vec_map.rs]: Is there a reason for this coercion? This type is known at compile time, surely the struct can simply store the closure by-value without the indirection of a function pointer?

Comment (Gankro) [inline src/libcollections/vec_map.rs]: Oh wait I'm thinking of something else. Nevermind.

File: src/libcollections/vec_map.rs [hunk 5]
```diff
@@ -709,6 +764,11 @@ impl<V> DoubleEndedIterator for IntoIter<V> {
     fn next_back(&mut self) -> Option<(uint, V)> { self.iter.next_back() }
 }
 
+#[stable]
+impl<V> ExactSizeIterator for IntoIter<V> {
+    fn len(&self) -> usize { self.len() }
```
Comment (Gankro) [inline src/libcollections/vec_map.rs]: This is just infinite recursion...?

File: src/libcollections/vec_map.rs [hunk 6]
```diff
@@ -673,6 +701,33 @@ pub struct IntoIter<V> {
...
+}
+
+#[unstable]
+impl<'a, V> DoubleEndedIterator for Drain<'a, V> {
+    fn next_back(&mut self) -> Option<(uint, V)> { self.iter.next_back() }
+}
+
+#[unstable]
+impl<'a, V> ExactSizeIterator for Drain<'a, V> {
+    fn len(&self) -> usize { self.len() }
```
Comment (Gankro) [inline src/libcollections/vec_map.rs]: Infinite recursion...?

**Comment by Gankro** [comment]
I would suggest against ExactSizeIterator; `len` is an O(n) operation for VecMap.

File: src/libcollections/vec_map.rs (hunk 5 above)
Comment (aochagavia) [inline src/libcollections/vec_map.rs]: Nice catch! It should be `self.iter.len()`... But self.iter doesn't implement `ExactSizeIterator` so I'll just remove it

**Comment by Gankro** [comment]
Needs tests

**Comment by aochagavia** [comment]
@Gankro I fixed the issues you pointed out and added a test

**Comment by Gankro** [comment]
Woah a run-pass test is way overkill. Just need some basic tests in the module that make sure the iterators produce correct output (and e.g. `drain` actually empties the parent). Check out VecMap's tests and other collections for examples (we're actually a bit inconsistent in how we test iterators).

**Comment by aochagavia** [comment]
@Gankro Sorry for the run-pass test :yum: ... I have added it as a normal test.

**Comment by Gankro** [comment]
;_;

**Comment by aochagavia** [comment]
Maybe it needs a longer string?

See here: https://github.com/rust-lang/rust/pull/21386

**Comment by Gankro** [comment]
It registered the last one

**Comment by Gankro** [comment]
Also I clearly nailed it on my first try.

Anyone who claims otherwise is clearly a liar?

**Comment by aochagavia** [comment]
Thanks for reviewing!

**Comment by Gankro** [comment]
Thanks for contributing! :confetti_ball:

**Pull request merged**